packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts

```diff
@@ -0,0 +1,228 @@
+    let snapshotOptions: ISnapshotOptions = {
+        deltas: 1,
+        channels: 1,
+        blobs: 2,
+    };
+
+    if (hostSnapshotFetchOptions) {
+        snapshotOptions = {...snapshotOptions, ...hostSnapshotFetchOptions};
+    }
+
+    const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+        odspLogger,
+        odspResolvedUrl,
+        getStorageToken,
+        true /* throwOnNullToken */,
+    );
+
+    const snapshotCacheEntry: ICacheEntry = {
+        type: snapshotKey,
+        key: "",
+        file: {
+            resolvedUrl: odspResolvedUrl,
+            docId: odspResolvedUrl.hashedDocumentId,
+        },
+    };
+
+    return getWithRetryForTokenRefresh(async (tokenFetchOptions) => {
+        // No need to ask cache twice - if first request was unsuccessful, cache unlikely to have data on second turn.
+        if (tokenFetchOptions.refresh) {
+            await fetchSnapshot(
+                snapshotUrl,
+                odspResolvedUrl,
+                snapshotOptions,
+                tokenFetchOptions,
+                storageTokenFetcher,
+                odspLogger,
+                persistedCache,
+                snapshotCacheEntry,
+            );
+        } else {
+            await PerformanceEvent.timedExecAsync(
+                odspLogger,
+                { eventName: "ObtainSnapshot" },
+                async (event: PerformanceEvent) => {
+                    const cachedSnapshot: ISnapshotCacheValue | undefined =
+                        await persistedCache.get(snapshotCacheEntry);
+                    const method = cachedSnapshot !== undefined ? "cache" : "network";
+
+                    if (cachedSnapshot === undefined) {
+                        await fetchSnapshot(
+                            snapshotUrl,
+                            odspResolvedUrl,
+                            snapshotOptions,
+                            tokenFetchOptions,
+                            storageTokenFetcher,
+                            odspLogger,
+                            persistedCache,
+                            snapshotCacheEntry,
+                        );
+                    }
+                    event.end({ method });
+                    return cachedSnapshot;
+                },
+            );
+        }
+        if (persistedCache.get(snapshotCacheEntry) !== undefined) {
+            return true;
+        }
+        return false;
+    }).catch((error) => {
+        return false;
+    });
+}
+
+async function fetchSnapshot(
+    snapshotUrl: string,
+    odspResolvedUrl: IOdspResolvedUrl,
+    snapshotOptions: ISnapshotOptions,
+    tokenFetchOptions: TokenFetchOptions,
+    getStorageToken: (options: TokenFetchOptions, name: string) => Promise<string | null>,
+    logger: TelemetryLogger,
+    persistedCache: IPersistedCache,
+    snapshotCacheEntry: ICacheEntry,
+): Promise<void> {
+    const storageToken = await getStorageToken(tokenFetchOptions, "TreesLatest");
```

Comment:
I'd move this call to the top of getWithRetryForTokenRefresh() callback and just pass token into here (assuming we always want to hit storage). Otherwise we can keep it here

Reply:
Done.